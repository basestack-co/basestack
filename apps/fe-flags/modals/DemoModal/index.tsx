import React, { useCallback, useMemo } from "react";
// Store
import { batch, useDispatch, useSelector } from "react-redux";
import { AppDispatch, RootState } from "store";
import { setDemoModalOpen } from "store/slices/modals";
// Utils
import Portal from "design-system/global/Portal";
// import dayjs from "dayjs";
// Locales
import { useIntl } from "react-intl";
import { demoMessages as messages } from "locales/messages/app";

const DemoModal = () => {
  const intl = useIntl();
  const dispatch = useDispatch<AppDispatch>();
  const isModalOpen = useSelector(
    (store: RootState) => store.modals.isDemoModalOpen
  );

  const onClose = useCallback(() => {
    if (isModalOpen) {
      batch(() => {
        dispatch(setDemoModalOpen(false));
      });
    }
  }, [isModalOpen, dispatch]);

  return useMemo(() => {
    if (!isModalOpen) return null;

    return (
      <Portal selector="#portal">
        <div>Demo portal</div>
        <p>intl: {intl.formatMessage(messages.demo)}</p>
        <button onClick={onClose}>Close Modal</button>
      </Portal>
    );
  }, [isModalOpen]);
};

export default DemoModal;
