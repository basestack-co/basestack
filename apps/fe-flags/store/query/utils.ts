import {
  MiddlewareAPI,
  isRejectedWithValue,
  Middleware,
} from "@reduxjs/toolkit";

export const baseUrl = `http://localhost:3000/api/v1`;

export const prepareHeaders = (headers: Headers) => {
  // const state = getState() as RootState;
  return headers;
};

export const rtkQueryErrorLogger: Middleware =
  (api: MiddlewareAPI) => (next) => (action) => {
    if (isRejectedWithValue(action)) {
      const {
        payload: { status },
      } = action;
      // Access forbidden
      if (status === 403) {
        console.log("rtkQueryErrorLogger Access forbidden");
      }
    }
    return next(action);
  };
