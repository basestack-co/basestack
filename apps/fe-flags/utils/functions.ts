// Utils
import isEmpty from "lodash.isempty";

export const isBrowser = typeof window !== "undefined";

export const isNavigator = typeof navigator !== "undefined";

/**
 * UTILS OPERATIONS
 */

export const validURL = (str: string) => {
  const pattern = new RegExp(
    "^(https?:\\/\\/)?" + // protocol
      "((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|" + // domain name
      "((\\d{1,3}\\.){3}\\d{1,3}))" + // OR ip (v4) address
      "(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*" + // port and path
      "(\\?[;&a-z\\d%_.~+=-]*)?" + // query string
      "(\\#[-a-z\\d_]*)?$",
    "i"
  );
  return pattern.test(str);
};

export const urlQueryBuilder = (urlParams: any) => {
  const esc = encodeURIComponent;
  return Object.keys(urlParams)
    .filter((x) => !isEmpty(urlParams[x]))
    .map((k) => esc(k) + "=" + esc(urlParams[k]))
    .join("&");
};

export function groupArrayOfObjects(list: Array<any>, key: string) {
  return list.reduce((rv, x) => {
    (rv[x[key]] = rv[x[key]] || []).push(x);
    return rv;
  }, {});
}

export function groupBy(xs, f) {
  return xs.reduce(
    (r, v, i, a, k = f(v)) => ((r[k] || (r[k] = [])).push(v), r),
    {}
  );
}

export function getBaseUrl() {
  if (typeof window !== "undefined") {
    return "";
  }
  // reference for vercel.com
  if (process.env.VERCEL_URL) {
    return `https://${process.env.VERCEL_URL}`;
  }

  // // reference for render.com
  if (process.env.RENDER_INTERNAL_HOSTNAME) {
    return `http://${process.env.RENDER_INTERNAL_HOSTNAME}:${process.env.PORT}`;
  }

  // assume localhost
  return `http://localhost:${process.env.PORT ?? 3000}`;
}
