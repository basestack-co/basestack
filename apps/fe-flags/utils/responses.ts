// MESSAGES

export const unauthorized = {
  code: 401,
  error: true,
  message: "You must be sign in to view the protected content on this page.",
};

export const methodNotAllowed = {
  code: 405,
  error: true,
  message: `The HTTP method is not supported at this route.`,
};

export const somethingWentWrong = "Something went wrong";

export const notAuthorizedActionProject =
  "You are not authorized to do this action in this project";

// ENVIRONMENTS

export const notAuthorizedCreateEnv =
  "You are not authorized to create an environment";

export const notAuthorizedUpdateEnv =
  "You are not authorized to update this environment";

export const notAuthorizedDeleteEnv =
  "You are not authorized to delete this environment";

// FLAGS

export const notAuthorizedCreateFlag =
  "You are not authorized to create an flag";

export const notAuthorizedUpdateFlag =
  "You are not authorized to update this flag";

export const notAuthorizedDeleteFlag =
  "You are not authorized to delete this flag";

// HISTORY

export const notAuthorizedCreateHistory =
  "You are not authorized to create an history";

export const notAuthorizedUpdateHistory =
  "You are not authorized to update this history";
