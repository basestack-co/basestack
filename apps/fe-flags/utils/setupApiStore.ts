import {
  AnyAction,
  combineReducers,
  configureStore,
  EnhancedStore,
  Middleware,
  Reducer,
} from "@reduxjs/toolkit";

export function setupApiStore<
  A extends {
    reducer: Reducer<any, any>;
    reducerPath: string;
    middleware: Middleware;
    util: { resetApiState(): any };
  },
  R extends Record<string, Reducer<any, any>> = Record<never, never>
>(api: A, extraReducers?: R): { api: any; store: EnhancedStore } {
  /*
   * Modified version of RTK Query's helper function:
   * https://github.com/reduxjs/redux-toolkit/blob/master/packages/toolkit/src/query/tests/helpers.tsx
   */
  const getStore = (): EnhancedStore =>
    configureStore({
      reducer: combineReducers({
        [api.reducerPath]: api.reducer,
        ...extraReducers,
      }),
      middleware: (gdm) =>
        gdm({ serializableCheck: false, immutableCheck: false }).concat(
          api.middleware
        ),
    });

  type StoreType = EnhancedStore<
    {
      api: ReturnType<A["reducer"]>;
    } & {
      [K in keyof R]: ReturnType<R[K]>;
    },
    AnyAction,
    ReturnType<typeof getStore> extends EnhancedStore<any, any, infer M>
      ? M
      : never
  >;

  const initialStore = getStore() as StoreType;
  const refObj = {
    api,
    store: initialStore,
  };
  const store = getStore() as StoreType;
  refObj.store = store;

  return refObj;
}
