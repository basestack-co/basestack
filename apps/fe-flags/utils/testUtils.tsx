import React from "react";
import { render, RenderOptions } from "@testing-library/react";
// Store
import { Provider } from "react-redux";
import { store } from "../store";
// Locales
import { IntlProvider } from "react-intl";
import { messages } from "../locales";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "design-system/theme";
// Auth
import { SessionProvider } from "next-auth/react";

interface AllProvidersProps {
  initialState?: {};
}

const AllProviders: React.FC<AllProvidersProps> = ({
  children,
  initialState = {},
}) => {
  return (
    <SessionProvider session={{ expires: null }}>
      <Provider store={{ ...store, ...initialState }}>
        <IntlProvider locale="en" messages={messages.en}>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </IntlProvider>
      </Provider>
    </SessionProvider>
  );
};

const renderWithAllProviders = (
  ui: React.ReactElement,
  initialState?: {},
  options?: Omit<RenderOptions, "wrapper">
) =>
  render(ui, {
    wrapper: (props) => <AllProviders initialState={initialState} {...props} />,
    ...options,
  });

const WithThemeProvider: React.FC = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: WithThemeProvider, ...options });

const WithReduxProvider: React.FC = ({ children }) => (
  <Provider store={store}>{children}</Provider>
);

const renderWithRedux = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: WithReduxProvider, ...options });

export { renderWithAllProviders, renderWithTheme, renderWithRedux };
