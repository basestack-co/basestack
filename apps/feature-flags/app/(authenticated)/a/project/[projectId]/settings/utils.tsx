// Utils
import { config, PlanTypeId, FlagsPlan, Product } from "@basestack/utils";
// Types
import { useRouter } from "next/navigation";
// Design System
import { CardVariant } from "@basestack/design-system";

export interface WithPlanCardProps {
  planId: PlanTypeId;
  feature: keyof FlagsPlan["features"];
  onClick: () => void;
  isLoading: boolean;
  isDisabled: boolean;
  partial?: boolean;
  router: ReturnType<typeof useRouter>;
  labels: {
    partial: string;
    all: string;
    upgrade: string;
    save: string;
    text: string;
  };
}

export const getWithPlanCardProps = ({
  planId,
  feature,
  onClick,
  isLoading,
  isDisabled,
  partial = true,
  router,
  labels,
}: WithPlanCardProps) => {
  const hasFeature = config.plans.hasPlanFeature(
    Product.FLAGS,
    planId,
    feature,
  );

  return !hasFeature
    ? {
        button: labels.upgrade,
        onClick: () => router.push("/a/user/tab/billing"),
        hasOverlay: true,
        variant: CardVariant.PRIMARY,
        label: partial ? labels.partial : labels.all,
      }
    : {
        button: labels.save,
        text: labels.text,
        onClick,
        isLoading,
        isDisabled,
      };
};
