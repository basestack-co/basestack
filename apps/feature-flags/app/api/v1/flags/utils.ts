// Utils
import {
  PlanTypeId,
  config as utilsConfig,
  config,
  isRefererValid,
  getValidWebsite,
  isValidIpAddress,
  RequestError,
} from "@basestack/utils";
// DB
import { prisma } from "server/db";
import { getProjectOnUser, productUrl } from "server/db/utils/project";
import { withUsageUpdate } from "server/db/utils/subscription";

const { hasFlagsPlanFeature } = utilsConfig.plans;

export const verifyRequest = async (
  key: string,
  referer: string,
  metadata: { ip: string | null },
) => {
  try {
    const projectKey = key.trim();

    const project = await getProjectOnUser(projectKey);

    if (!project) {
      console.info(`"Error: No project found with the ID in the request`, {
        product: "Feature Flags",
        referer,
      });

      throw new RequestError({
        code: 404,
        url: productUrl,
        message: "Error: No project found",
      });
    }

    const planId = project.usage.planId as PlanTypeId;

    if (
      !!project.websites &&
      hasFlagsPlanFeature(planId, "hasWebsites") &&
      isRefererValid(referer)
    ) {
      const isValid = getValidWebsite(referer, project.websites);

      if (!isValid) {
        console.info(
          `"Error: The website is not allowed to fetch feature flags`,
          {
            product: "Feature Flags",
            referer,
          },
        );

        throw new RequestError({
          code: 403,
          url: productUrl,
          message: "Unauthorized request",
        });
      }
    }

    if (
      !!project.blockIpAddresses &&
      metadata?.ip &&
      hasFlagsPlanFeature(planId, "hasBlockIPs")
    ) {
      const blockIpAddresses = project.blockIpAddresses
        .split(",")
        .map((ip) => ip.trim())
        .filter(Boolean)
        .filter(isValidIpAddress);

      const clientIp = metadata.ip.trim();

      if (clientIp && isValidIpAddress(clientIp)) {
        const isBlocked = blockIpAddresses.some((ip) => ip === clientIp);

        if (isBlocked) {
          console.info(
            `Error: Your IP address is blocked from fetching feature flags`,
            {
              product: "Feature Flags",
              referer,
            },
          );

          throw new RequestError({
            code: 403,
            url: productUrl,
            message: "Access denied",
          });
        }
      }
    }

    const limit = config.plans.getFlagsLimitByKey(planId, "apiRequests");

    if (project.usage.apiRequests >= limit) {
      console.info(
        `Error: Plan limit exceeded for API requests. Please consider upgrading.`,
        {
          product: "Feature Flags",
          referer,
        },
      );

      throw new RequestError({
        code: 429,
        url: productUrl,
        message: "Rate limit exceeded",
      });
    }

    await withUsageUpdate(prisma, project.userId, "apiRequests", "increment");

    return true;
  } catch (error) {
    if (!(error instanceof RequestError)) {
      console.error("Unexpected error in verifyRequest:", error);
      throw new RequestError({
        code: 500,
        url: productUrl,
        message: "An internal error occurred",
      });
    }

    throw error;
  }
};
