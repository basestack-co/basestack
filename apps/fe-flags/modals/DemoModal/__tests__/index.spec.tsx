import React from "react";
import { renderWithAllProviders } from "utils/testUtils";
import DemoModal from "..";

describe("DemoModal tests", () => {
  test("render DemoModal correctly", () => {
    const { asFragment } = renderWithAllProviders(<DemoModal />);
    expect(asFragment()).toMatchSnapshot();
  });
});
