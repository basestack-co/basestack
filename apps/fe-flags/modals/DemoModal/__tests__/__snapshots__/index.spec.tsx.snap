// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DemoModal tests render DemoModal correctly 1`] = `<DocumentFragment />`;
